File: 242-valid-anagram/solution.py
# Valid Anagram
# https://leetcode.com/problems/valid-anagram/

class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        freq = {}

        for char in s:
            if char not in freq:
                freq[char] = 0
            freq[char] += 1

        for char in t:
            if char not in freq:
                return False
            freq[char] -= 1
            if not freq[char]:
                del freq[char]

        return len(freq) == 0

File: 242-valid-anagram/solution.js
// Valid Anagram
// https://leetcode.com/problems/valid-anagram/

/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function (s, t) {
  if (s.length !== t.length) return false;

  const freq = new Map();

  for (let char of s) {
    freq.set(char, (freq.get(char) || 0) + 1);
  }

  for (let char of t) {
    if (!freq.has(char)) return false;
    freq.set(char, freq.get(char) - 1);
    if (freq.get(char) === 0) freq.delete(char);
  }

  return freq.size === 0;
};

// Time: O(n)
// Space: O(1) ~ 26 characters
